File: union_find/java/WQUPC.java
import java.util.*;

public class WQUPC {

  private int[] parent;
  private int[] size;
  private int count;

  public WQUPC(int N) {
    int i;
    count = N;
    parent = new int[N];
    size = new int[N];
    for(i = 0; i < N; i++) {
      parent[i] = i;
      size[i] = 1;
    }
  }

  private int root(int i) {
    while(i != parent[i]) {
      parent[i] = parent[parent[i]];
      i = parent[i];
    }
    return i;
  }

  public boolean connected(int p, int q) {
    return root(p) == root(q);
  }

  public void union(int p, int q) {
    if(connected(p, q)) return;
    int pRoot = root(p);
    int qRoot = root(q);

    if(size[pRoot] > size[qRoot]) {
      parent[qRoot] = pRoot;
      size[pRoot] += size[qRoot];
    } else {
      parent[pRoot] = qRoot;
      size[qRoot] += size[pRoot];
    }
    count--;

    System.out.println(Arrays.toString(parent));
  }
}

File: union_find/javascript/WQUPC.js
class WQUPC {
  constructor(N) {
    var i
    this.count = N
    this.parent = []
    this.size = []
    for(i = 0; i < N; i++) {
      this.parent.push(i)
      this.size.push(1)
    }
  }

  root(i) {
    while(i != this.parent[i]) {
      this.parent[i] = this.parent[this.parent[i]]
      i = this.parent[i]
    }
    return i
  }

  connected(p, q) {
    return this.root(p) == this.root(q)
  }

  union(p, q) {
    if(this.connected(p, q)) return;
    var pRoot = this.root(p)
    var qRoot = this.root(q)

    if(this.size[pRoot] > this.size[qRoot]) {
      this.parent[qRoot] = pRoot;
      this.size[pRoot] += this.size[qRoot]
    } else {
      this.parent[pRoot] = qRoot;
      this.size[qRoot] += this.size[pRoot];
    }
    this.count--;

    console.log(this.parent)
  }
}
